File: scripts/verify_phase_witness.py
"""Regenerate candidate checks with the original solver (requires GNU Octave)."""

import subprocess
from pathlib import Path

import numpy as np
from scipy.io import loadmat, savemat

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "research/results"
for name in ["probe", "tight"]:
    data = np.load(OUT / f"phase-ambiguity-{name}.npz")
    savemat(OUT / f"phase-ambiguity-{name}-input.mat", {"x": data["candidate_x"]}, oned_as="column")
subprocess.run(
    ["octave", "--no-gui", "--quiet", "scripts/verify_phase_witness.m"], cwd=ROOT, check=True
)
for name in ["probe", "tight"]:
    data = np.load(OUT / f"phase-ambiguity-{name}.npz")
    ref = loadmat(OUT / f"phase-ambiguity-{name}-reference.mat", simplify_cells=True)
    delta = float(np.max(abs(np.column_stack([ref["a"], ref["b"]]) - data["candidate_amplitude"])))
    if delta >= 1e-12:
        raise RuntimeError(f"{name}: original-code mismatch {delta:g}")
    print(f"{name}: original-code max complex-amplitude difference {delta:.3g}")

File: scripts/verify_phase_witness.m
addpath(genpath('research/sources/sherwin'));
m=load('research/results/sherwin-input-131.mat');
for field={'N_ML','n_cap','n_per','n_contam','n_etch'}
    m.(field{1})=double(m.(field{1}));
end
for name={'probe','tight'}
    d=load(['research/results/phase-ambiguity-' name{1} '-input.mat']);x=d.x;
    [a,b]=calcMLAbsEtchFresnel(m.N_ML,m.n_cap,m.n_per,m.n_contam,m.n_etch, ...
        x(m.ind_struct.thick),m.amu,m.rho_nom,x(m.ind_struct.rough), ...
        m.f0f1_elements_r,reshape(x(m.ind_struct.composition),length(m.amu),[]), ...
        m.lambdaTheta,m.pol,1,x(m.ind_struct.substrate_roughness));
    save('-mat7-binary',['research/results/phase-ambiguity-' name{1} '-reference.mat'],'a','b');
end
